File: src/uk/co/ben_gibson/git/link/UI/Action/Vcs/VcsLogAction.java
package uk.co.ben_gibson.git.link.UI.Action.Vcs;

import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.project.Project;
import com.intellij.vcs.log.VcsFullCommitDetails;
import com.intellij.vcs.log.VcsLog;
import com.intellij.vcs.log.VcsLogDataKeys;
import uk.co.ben_gibson.git.link.Container;
import uk.co.ben_gibson.git.link.Git.Commit;
import uk.co.ben_gibson.git.link.Url.Handler.UrlHandler;
import uk.co.ben_gibson.git.link.UI.Action.Action;

import java.util.List;

/**
 * An action triggered from an VCS log toolbar.
 */
abstract class VcsLogAction extends Action
{
    abstract UrlHandler urlHandler();


    public void actionPerformed(Project project, AnActionEvent event)
    {
        VcsLog vcsLog = event.getData(VcsLogDataKeys.VCS_LOG);

        if (vcsLog == null) {
            return;
        }

        VcsFullCommitDetails vcsCommit = vcsLog.getSelectedDetails().get(0);
        Commit commit = new Commit(vcsCommit.toString());

        this.getManager().handleCommit(this.urlHandler(), project, commit, vcsCommit.getRoot());
    }


    protected boolean shouldActionBeEnabled(AnActionEvent event)
    {
        VcsLog log = event.getData(VcsLogDataKeys.VCS_LOG);

        if (log == null) {
            return false;
        }

        List<VcsFullCommitDetails> commits = log.getSelectedDetails();

        return commits.size() == 1;
    }
}

File: src/main/kotlin/uk/co/ben_gibson/git/link/ui/actions/vcslog/VcsLogAction.kt
package uk.co.ben_gibson.git.link.ui.actions.vcslog

import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.project.Project
import com.intellij.vcs.log.VcsLogDataKeys.VCS_LOG_COMMIT_SELECTION
import uk.co.ben_gibson.git.link.Context
import uk.co.ben_gibson.git.link.git.Commit
import uk.co.ben_gibson.git.link.ui.actions.Action

abstract class VcsLogAction(type: Type) : Action(type) {
    override fun buildContext(project: Project, event: AnActionEvent): Context? {
        val vcsCommit = event.getData(VCS_LOG_COMMIT_SELECTION)?.cachedFullDetails?.firstOrNull() ?: return null

        return Context.Commit(vcsCommit.root, Commit(vcsCommit.id.asString()))
    }

    override fun shouldBeEnabled(event: AnActionEvent): Boolean {
        return event.getData(VCS_LOG_COMMIT_SELECTION)?.cachedFullDetails?.size == 1
    }
}
